## Week 4, Lab 1: Predicting Left-Handedness from Psychological Factors
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

We'll walk through a full data science problem in this lab. 

---
## Step 1: Define The Problem.

You're currently a data scientist working at a university. A professor of psychology is attempting to study the relationship between personalities and left-handedness. They have tasked you with gathering evidence so that they may publish.

Specifically, the professor says "I need to prove that left-handedness is caused by some personality trait. Go find that personality trait and the data to back it up."

As a data scientist, you know that any real data science problem must be **specific** and **conclusively answerable**. For example:
- Bad data science problem: "What is the link between obesity and blood pressure?"
    - This is vague and is not conclusively answerable. That is, two people might look at the conclusion and one may say "Sure, the problem has been answered!" and the other may say "The problem has not yet been answered."
- Good data science problem: "Does an association exist between obesity and blood pressure?"
    - This is more specific and is conclusively answerable. The problem specifically is asking for a "Yes" or "No" answer. Based on that, two independent people should both be able to say either "Yes, the problem has been answered" or "No, the problem has not yet been answered."
- Excellent data science problem: "As obesity increases, how does blood pressure change?"
    - This is very specific and is conclusively answerable. The problem specifically seeks to understand the effect of one variable on the other.

### 1. In the context of the left-handedness and personality example, what are three specific and conclusively answerable problems that you could answer using data science? 

> You might find it helpful to check out the codebook in the repo for some inspiration.

Answer:
- Does an association exist between personality and being left-handed?
- Does an association exist between emotional/mental health and being left-handed?
- Does an association exist between intelligence and being left-handed?

---
## Step 2: Obtain the data.

### 2. Read in the file titled "data.csv."
> Hint: Despite being saved as a .csv file, you won't be able to simply `pd.read_csv()` this data!

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('./data.csv', sep='\t')
df.head()

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  engnat  \
0   4   1   5   1   5   1   5   1   4    1  ...       US           2       1   
1   1   5   1   4   2   5   5   4   1    5  ...       CA           2       1   
2   1   2   1   1   5   4   3   2   1    4  ...       NL           2       2   
3   1   4   1   5   1   4   5   4   3    5  ...       US           2       1   
4   5   1   5   1   5   1   5   1   3    1  ...       US           2       1   

   age  education  gender  orientation  race  religion  hand  
0   22          3       1            1     3         2     3  
1   14          1       2            2     6         1     1  
2   30          4       1            1     1         1     2  
3   18          2       2            5     3         2     2  
4   22          3       1            1     3         2     3  

[5 rows x 56 columns]

### 3. Suppose that, instead of us giving you this data in a file, you were actually conducting a survey to gather this data yourself. From an ethics/privacy point of view, what are three things you might consider when attempting to gather this data?
> When working with sensitive data like sexual orientation or gender identity, we need to consider how this data could be used if it fell into the wrong hands!

Answer:
- keep participant confidentiality or conduct anonymously
- inform participant of the problem statement and how their input will be used, so as to ensure that they are comfortable with sharing
- ensure that survey question is structured in a way that personal data won't be interpreted differently

---
## Step 3: Explore the data.

### 4. Conduct exploratory data analysis on this dataset.
> If you haven't already, be sure to check out the codebook in the repo, as that will help in your EDA process.

In [3]:
df.head()

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  engnat  \
0   4   1   5   1   5   1   5   1   4    1  ...       US           2       1   
1   1   5   1   4   2   5   5   4   1    5  ...       CA           2       1   
2   1   2   1   1   5   4   3   2   1    4  ...       NL           2       2   
3   1   4   1   5   1   4   5   4   3    5  ...       US           2       1   
4   5   1   5   1   5   1   5   1   3    1  ...       US           2       1   

   age  education  gender  orientation  race  religion  hand  
0   22          3       1            1     3         2     3  
1   14          1       2            2     6         1     1  
2   30          4       1            1     1         1     2  
3   18          2       2            5     3         2     2  
4   22          3       1            1     3         2     3  

[5 rows x 56 columns]

In [4]:
df.isnull().sum()

Q1             0
Q2             0
Q3             0
Q4             0
Q5             0
Q6             0
Q7             0
Q8             0
Q9             0
Q10            0
Q11            0
Q12            0
Q13            0
Q14            0
Q15            0
Q16            0
Q17            0
Q18            0
Q19            0
Q20            0
Q21            0
Q22            0
Q23            0
Q24            0
Q25            0
Q26            0
Q27            0
Q28            0
Q29            0
Q30            0
Q31            0
Q32            0
Q33            0
Q34            0
Q35            0
Q36            0
Q37            0
Q38            0
Q39            0
Q40            0
Q41            0
Q42            0
Q43            0
Q44            0
introelapse    0
testelapse     0
country        0
fromgoogle     0
engnat         0
age            0
education      0
gender         0
orientation    0
race           0
religion       0
hand           0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4184 entries, 0 to 4183
Data columns (total 56 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Q1           4184 non-null   int64 
 1   Q2           4184 non-null   int64 
 2   Q3           4184 non-null   int64 
 3   Q4           4184 non-null   int64 
 4   Q5           4184 non-null   int64 
 5   Q6           4184 non-null   int64 
 6   Q7           4184 non-null   int64 
 7   Q8           4184 non-null   int64 
 8   Q9           4184 non-null   int64 
 9   Q10          4184 non-null   int64 
 10  Q11          4184 non-null   int64 
 11  Q12          4184 non-null   int64 
 12  Q13          4184 non-null   int64 
 13  Q14          4184 non-null   int64 
 14  Q15          4184 non-null   int64 
 15  Q16          4184 non-null   int64 
 16  Q17          4184 non-null   int64 
 17  Q18          4184 non-null   int64 
 18  Q19          4184 non-null   int64 
 19  Q20          4184 non-null 

In [6]:
df.iloc[:, 0:44].describe()

Q1           Q2           Q3           Q4           Q5  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000   
mean      1.962715     3.829589     2.846558     3.186902     2.865440   
std       1.360291     1.551683     1.664804     1.476879     1.545798   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     3.000000     1.000000     2.000000     1.000000   
50%       1.000000     5.000000     3.000000     3.000000     3.000000   
75%       3.000000     5.000000     5.000000     5.000000     4.000000   
max       5.000000     5.000000     5.000000     5.000000     5.000000   

                Q6           Q7           Q8           Q9          Q10  ...  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  ...   
mean      3.672084     3.216539     3.184512     2.761233     3.522945  ...   
std       1.342238     1.490733     1.387382     1.511805     1.242890  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       3.000000     2.000000     2.000000     1.000000     3.000000  ...   
50%       4.000000     3.000000     3.000000     3.000000     4.000000  ...   
75%       5.000000     5.000000     4.000000     4.000000     5.000000  ...   
max       5.000000     5.000000     5.000000     5.000000     5.000000  ...   

               Q35          Q36          Q37          Q38          Q39  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000   
mean      3.074092     2.610660     3.465344     2.798757     2.569312   
std       1.546400     1.409707     1.521460     1.413584     1.621772   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       1.000000     1.000000     2.000000     1.000000     1.000000   
50%       3.000000     2.000000     4.000000     3.000000     2.000000   
75%       5.000000     4.000000     5.000000     4.000000     4.000000   
max       5.000000     5.000000     5.000000     5.000000     5.000000   

               Q40          Q41          Q42          Q43          Q44  
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000  
mean      2.984226     3.385277     2.704828     2.676386     2.736616  
std       1.483752     1.423055     1.544345     1.523097     1.471845  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       2.000000     2.000000     1.000000     1.000000     1.000000  
50%       3.000000     4.000000     3.000000     3.000000     3.000000  
75%       4.000000     5.000000     4.000000     4.000000     4.000000  
max       5.000000     5.000000     5.000000     5.000000     5.000000  

[8 rows x 44 columns]

In [7]:
df.iloc[:, 44:].describe()

introelapse     testelapse   fromgoogle       engnat           age  \
count    4184.000000    4184.000000  4184.000000  4184.000000   4184.000000   
mean      347.808556     479.994503     1.576243     1.239962     30.370698   
std      5908.901681    3142.178542     0.494212     0.440882    367.201726   
min         1.000000       7.000000     1.000000     0.000000     13.000000   
25%         6.000000     186.000000     1.000000     1.000000     18.000000   
50%        12.000000     242.000000     2.000000     1.000000     21.000000   
75%        35.000000     324.250000     2.000000     1.000000     27.000000   
max    252063.000000  119834.000000     2.000000     2.000000  23763.000000   

         education       gender  orientation         race     religion  \
count  4184.000000  4184.000000  4184.000000  4184.000000  4184.000000   
mean      2.317878     1.654398     1.833413     5.013623     2.394359   
std       0.874264     0.640915     1.303454     1.970996     2.184164   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       2.000000     1.000000     1.000000     5.000000     1.000000   
50%       2.000000     2.000000     1.000000     6.000000     2.000000   
75%       3.000000     2.000000     2.000000     6.000000     2.000000   
max       4.000000     3.000000     5.000000     7.000000     7.000000   

              hand  
count  4184.000000  
mean      1.190966  
std       0.495357  
min       0.000000  
25%       1.000000  
50%       1.000000  
75%       1.000000  
max       3.000000

Analysis:
- max 'age' is not possible
- min 'testslapse' is not short
- min 'hand' 0 is not valid data

In [8]:
df['hand'].value_counts()

1    3542
2     452
3     179
0      11
Name: hand, dtype: int64

In [9]:
df = df[df.hand != 0]

In [10]:
df['hand'].value_counts()

1    3542
2     452
3     179
Name: hand, dtype: int64

In [11]:
df.iloc[:, 44:].sort_values('age',ascending=False).head()

introelapse  testelapse country  fromgoogle  engnat  age  education  \
2137           37         352      US           2       1  409          2   
2075          107         143      US           2       1  123          1   
2101           16         236      US           1       1   86          3   
1736            6         257      US           2       1   86          3   
3436           25         264      US           2       1   85          4   

      gender  orientation  race  religion  hand  
2137       1            5     6         1     1  
2075       1            5     7         7     3  
2101       1            3     6         2     1  
1736       1            3     6         2     1  
3436       1            1     6         1     1

In [12]:
df = df[df.age < 121]

In [13]:
df.iloc[:, 44:].sort_values('age',ascending=False).head()

introelapse  testelapse country  fromgoogle  engnat  age  education  \
1736            6         257      US           2       1   86          3   
2101           16         236      US           1       1   86          3   
3436           25         264      US           2       1   85          4   
1909           11         386      US           2       1   78          2   
3182           51         307      CA           1       1   77          4   

      gender  orientation  race  religion  hand  
1736       1            3     6         2     1  
2101       1            3     6         2     1  
3436       1            1     6         1     1  
1909       1            1     6         0     1  
3182       1            1     6         2     1

In [14]:
df.iloc[:, 44:].sort_values('testelapse').head(10)

introelapse  testelapse country  fromgoogle  engnat  age  education  \
3079            8           7      IN           2       1   35          4   
3414            4          14      US           1       1   26          3   
3788            5          41      US           1       1   20          2   
1868            2          48      US           1       1   21          3   
3134           37          49      US           2       1   15          1   
3000            2          50      US           1       1   23          3   
1775            5          59      US           1       1   16          2   
3839            5          67      GB           2       1   44          3   
2638           50          68      ID           1       1   32          3   
3931           28          70      DZ           2       1   14          1   

      gender  orientation  race  religion  hand  
3079       1            2     2         5     1  
3414       1            1     6         1     2  
3788       2            1     1         2     1  
1868       2            1     6         2     1  
3134       2            1     3         2     1  
3000       1            1     7         1     1  
1775       1            1     6         0     2  
3839       1            1     6         1     1  
2638       1            2     2         2     1  
3931       2            0     6         3     1

In [15]:
df = df[df.testelapse > 40]

In [16]:
df.iloc[:, 44:].sort_values('testelapse').head(10)

introelapse  testelapse country  fromgoogle  engnat  age  education  \
3788            5          41      US           1       1   20          2   
1868            2          48      US           1       1   21          3   
3134           37          49      US           2       1   15          1   
3000            2          50      US           1       1   23          3   
1775            5          59      US           1       1   16          2   
3839            5          67      GB           2       1   44          3   
2638           50          68      ID           1       1   32          3   
3931           28          70      DZ           2       1   14          1   
1817            5          78      US           1       1   21          3   
1970            4          90      US           2       1   17          2   

      gender  orientation  race  religion  hand  
3788       2            1     1         2     1  
1868       2            1     6         2     1  
3134       2            1     3         2     1  
3000       1            1     7         1     1  
1775       1            1     6         0     2  
3839       1            1     6         1     1  
2638       1            2     2         2     1  
3931       2            0     6         3     1  
1817       1            1     7         2     2  
1970       2            2     6         1     1

In [17]:
df['engnat'].value_counts()

1    3119
2    1025
0      25
Name: engnat, dtype: int64

---
## Step 4: Model the data.

### 5. Suppose I wanted to use Q1 - Q44 to predict whether or not the person is left-handed. Would this be a classification or regression problem? Why?

Answer: Classification. We are categorizing our prediction to be either left-handed or right-handed.

### 6. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed based on their responses to Q1 - Q44. Before doing that, however, you remember that it is often a good idea to standardize your variables. In general, why would we standardize our variables? Give an example of when we would standardize our variables.

Answer: If the scale of features is very different then normalization is required. This is because the distance calculation done in KNN uses feature values. When the one feature values are large than other, that feature will dominate the distance hence the outcome of the KNN.

### 7. Give an example of when we might not standardize our variables.

Answer: if the variables are proportionately scaled

### 8. Based on your answers to 6 and 7, do you think we should standardize our predictor variables in this case? Why or why not?

Answer: No, as our values for survey questions are from 1 to 5. Unless we use variable that are not of the same range, like 'age', 'introelapse' or 'testelapse'.

### 9. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed. What munging/cleaning do we need to do to our $y$ variable in order to explicitly answer this question? Do it.

Answer: The output of y variable should be True or 1 for left-handed and False or 0 for right-handed. We can use one-hot-encoding to create out new y variable column.

|Feature|Question|Description|
|---|---|---|
|hand|"What hand do you use to write with?"|1=Right, 2=Left, 3=Both|

In [18]:
df2 = pd.get_dummies(df,columns=['hand'])

In [19]:
df2['left'] = df2['hand_2'] + df2['hand_3']

In [20]:
df2.drop(columns = ['hand_1', 'hand_2', 'hand_3'], inplace=True)
df2.head()

Q1  Q2  Q3  Q4  Q5  Q6  Q7  Q8  Q9  Q10  ...  country  fromgoogle  engnat  \
0   4   1   5   1   5   1   5   1   4    1  ...       US           2       1   
1   1   5   1   4   2   5   5   4   1    5  ...       CA           2       1   
2   1   2   1   1   5   4   3   2   1    4  ...       NL           2       2   
3   1   4   1   5   1   4   5   4   3    5  ...       US           2       1   
4   5   1   5   1   5   1   5   1   3    1  ...       US           2       1   

   age  education  gender  orientation  race  religion  left  
0   22          3       1            1     3         2     1  
1   14          1       2            2     6         1     0  
2   30          4       1            1     1         1     1  
3   18          2       2            5     3         2     1  
4   22          3       1            1     3         2     1  

[5 rows x 56 columns]

In [21]:
df2['left'].value_counts()

0    3540
1     629
Name: left, dtype: int64

### 10. The professor for whom you work suggests that you set $k = 4$. In this specific case, why might this be a bad idea?

Answer: It might result in overfitting

### 11. Let's *(finally)* use $k$-nearest neighbors to predict whether or not a person is left-handed!

> Be sure to create a train/test split with your data!

> Create four separate models, one with $k = 3$, one with $k = 5$, one with $k = 15$, and one with $k = 25$.

> Instantiate and fit your models.

In [22]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

%matplotlib inline

In [23]:
# X1 = df2[['Q2','Q4','Q8','Q12','Q16','Q18','Q22','Q24','Q26','Q30','Q32','Q36','Q38','Q40','Q42','Q44','gender']]
# X2 = df2[['Q3','Q6','Q9','Q10','Q11','Q14','Q15','Q19','Q20','Q21','Q23','Q27','Q28','Q29','Q33','Q34','Q39','Q43','gender']]
# X3 = df2[['Q3','Q5','Q7','Q13','Q17','Q25','Q31','Q35','Q37','engnat','education','gender']]
X = df2.drop(columns=['introelapse','testelapse','country','fromgoogle','age','orientation','race','religion','left'])
y = df2['left']

In [24]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify = y)

In [25]:
ss = StandardScaler()
X_train_sc = ss.fit_transform(X_train)
X_test_sc = ss.transform(X_test)

### n_neighbors = 3

In [26]:
knn3 = KNeighborsClassifier(n_neighbors=3)
knn3.fit(X_train_sc, y_train)
knn3.score(X_train_sc, y_train)

0.8752399232245681

In [27]:
knn3.score(X_test_sc, y_test)

0.8168744007670182

In [28]:
cross_val_score(knn3, X_train_sc, y_train, cv=5).mean()

0.8035859424920128

### n_neighbors = 5

In [29]:
knn5 = KNeighborsClassifier(n_neighbors=5)
knn5.fit(X_train_sc, y_train)
knn5.score(X_train_sc, y_train)

0.8550863723608445

In [30]:
knn5.score(X_test_sc, y_test)

0.8331735378715245

In [31]:
cross_val_score(knn5, X_train_sc, y_train, cv=5).mean()

0.8269392971246006

### n_neighbors = 15

In [32]:
knn15 = KNeighborsClassifier(n_neighbors=15)
knn15.fit(X_train_sc, y_train)
knn15.score(X_train_sc, y_train)

0.8493282149712092

In [33]:
knn15.score(X_test_sc, y_test)

0.8494726749760306

In [34]:
cross_val_score(knn15, X_train_sc, y_train, cv=5).mean()

0.8483690734824281

### n_neighbors = 25

In [35]:
knn25 = KNeighborsClassifier(n_neighbors=25)
knn25.fit(X_train_sc, y_train)
knn25.score(X_train_sc, y_train)

0.8490083173384517

In [36]:
knn25.score(X_test_sc, y_test)

0.8494726749760306

In [37]:
cross_val_score(knn5, X_train_sc, y_train, cv=5).mean()

0.8269392971246006

Being good data scientists, we know that we might not run just one type of model. We might run many different models and see which is best.

### 12. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, let's check the [documentation for logistic regression in sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Is there default regularization? If so, what is it? If not, how do you know?

Answer: Regularization is applied by default. It is shown in the documentation 'penalty{‘l1’, ‘l2’, ‘elasticnet’, ‘none’}, default=’l2’'. L2 is Ridge Regression.

### 13. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, should we standardize our features?

Answer: Yes

### 14. Let's use logistic regression to predict whether or not the person is left-handed.


> Be sure to use the same train/test split with your data as with your $k$-NN model above!

> Create four separate models, one with LASSO and $\alpha = 1$, one with LASSO and $\alpha = 10$, one with Ridge and $\alpha = 1$, and one with Ridge and $\alpha = 10$. *(Hint: Be careful with how you specify $\alpha$ in your model!)*

> Instantiate and fit your models.

In [38]:
logreg_l1_1 = LogisticRegression(penalty = 'l1', C = 1.0, solver = 'saga')
logreg_l1_1.fit(X_train_sc, y_train)
logreg_l1_1.score(X_train_sc, y_train)

0.8490083173384517

In [39]:
logreg_l1_1.score(X_test_sc, y_test)

0.8494726749760306

In [40]:
cross_val_score(logreg_l1_1, X_train_sc, y_train, cv=5).mean()

0.8490085623003196

In [41]:
logreg_l1_10 = LogisticRegression(penalty = 'l1', C = 0.1, solver = 'saga')
logreg_l1_10.fit(X_train_sc, y_train)
logreg_l1_10.score(X_train_sc, y_train)

0.8490083173384517

In [42]:
logreg_l1_10.score(X_test_sc, y_test)

0.8494726749760306

In [43]:
cross_val_score(logreg_l1_10, X_train_sc, y_train, cv=5).mean()

0.8490085623003196

In [44]:
logreg_l2_1 = LogisticRegression(C = 1.0)
logreg_l2_1.fit(X_train_sc, y_train)
logreg_l2_1.score(X_train_sc, y_train)

0.8490083173384517

In [45]:
logreg_l2_1.score(X_test_sc, y_test)

0.8494726749760306

In [46]:
cross_val_score(logreg_l2_1, X_train_sc, y_train, cv=5).mean()

0.8490085623003196

In [47]:
logreg_l2_10 = LogisticRegression(C = 0.1)
logreg_l2_10.fit(X_train_sc, y_train)
logreg_l2_10.score(X_train_sc, y_train)

0.8490083173384517

In [48]:
logreg_l2_10.score(X_test_sc, y_test)

0.8494726749760306

In [49]:
cross_val_score(logreg_l2_10, X_train_sc, y_train, cv=5).mean()

0.8490085623003196

---
## Step 5: Evaluate the model(s).

### 15. Before calculating any score on your data, take a step back. Think about your $X$ variable and your $Y$ variable. Do you think your $X$ variables will do a good job of predicting your $Y$ variable? Why or why not? What impact do you think this will have on your scores?

Answer: I think the X variables qns categories are all mixed up. I do feel like it needed to be in more sepcific grouping. Catergoring the questions before testing. 

### 16. Using accuracy as your metric, evaluate all eight of your models on both the training and testing sets. Put your scores below. (If you want to be fancy and generate a table in Markdown, there's a [Markdown table generator site linked here](https://www.tablesgenerator.com/markdown_tables#).)
- Note: Your answers here might look a little weird. You didn't do anything wrong; that's to be expected!

|Model|Parameters|Train Score|Test Score|CV Score|
|:---|:-------------:|------:|------:|------:|
|k-Nearest Neighbors|n_neighbour = 3|0.8752399232245681|0.8168744007670182|0.8045516916523306|
|k-Nearest Neighbors|n_neighbour = 5|0.8550863723608445|0.8331735378715245|0.8307866388137952|
|k-Nearest Neighbors|n_neighbour = 15|0.8493282149712092|0.8494726749760306|0.848049275006144|
|k-Nearest Neighbors|n_neighbour = 25|0.8490083173384517|0.8494726749760306|0.8307866388137952|
|Logistic Regression|Lasso, 𝛼 = 1|0.8490083173384517|0.8494726749760306|0.8490085623003196|
|Logistic Regression|Lasso, 𝛼 = 10|0.8490083173384517|0.8494726749760306|0.8490085623003196|
|Logistic Regression|Ridge, 𝛼 = 1|0.8490083173384517|0.8494726749760306|0.8490085623003196|
|Logistic Regression|Ridge, 𝛼 = 10|0.8490083173384517|0.8494726749760306|0.8490085623003196|

### 17. In which of your $k$-NN models is there evidence of overfitting? How do you know?

Answer: The one with n_neighbour = 3 and 5, as the training score is much higher than the test score.

### 18. Broadly speaking, how does the value of $k$ in $k$-NN affect the bias-variance tradeoff? (i.e. As $k$ increases, how are bias and variance affected?)

Answer: When we increase K, the training error will increase (increase bias), but the test error may decrease at the same time (decrease variance).

### 19. If you have a $k$-NN model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer: 
- try using grid search to find the best param for k = n
- get more data row
- select more relevent qns for features

### 20. In which of your logistic regression models is there evidence of overfitting? How do you know?

Answer: All models have the same result. All their scores are quite close, thus no evidence of overfitting.

### 21. Broadly speaking, how does the value of $C$ in logistic regression affect the bias-variance tradeoff? (i.e. As $C$ increases, how are bias and variance affected?)

Answer: When we increase C, bias will increase, thus underfitting

### 22. For your logistic regression models, play around with the regularization hyperparameter, $C$. As you vary $C$, what happens to the fit and coefficients in the model? What do you think this means in the context of this specific problem?

In [50]:
logreg_l1_1.coef_

array([[ 0.02307529, -0.00513949,  0.13479876,  0.02357104,  0.01986972,
        -0.07439096,  0.03440509, -0.15386977, -0.12997462, -0.00688046,
         0.        ,  0.03589063, -0.00486613, -0.06720914, -0.01604056,
         0.06244799,  0.07461695, -0.0575968 ,  0.        , -0.04185905,
        -0.10169811, -0.10451301, -0.11517372, -0.01078743,  0.08977114,
         0.13697412,  0.16718803,  0.07232743,  0.03853853, -0.00344179,
        -0.01765597, -0.02593994,  0.07596326,  0.01336693,  0.17832208,
        -0.05914246,  0.02044956,  0.0341439 , -0.15278653, -0.00563056,
        -0.02276881, -0.06224102, -0.23582888,  0.06533168, -0.10293685,
        -0.004725  , -0.00154582]])

In [51]:
logreg_l1_10.coef_

array([[ 0.00928028,  0.        ,  0.10236595,  0.        ,  0.        ,
        -0.05933317,  0.00739564, -0.07860044, -0.07583559,  0.        ,
         0.        ,  0.        ,  0.        , -0.03509974,  0.        ,
         0.02869786,  0.04625513, -0.04057732,  0.        , -0.02714637,
        -0.02545408, -0.07206083, -0.09368283,  0.        ,  0.06888071,
         0.10497343,  0.        ,  0.03643097,  0.00764293,  0.        ,
         0.        , -0.00237005,  0.04004584,  0.        ,  0.14691613,
        -0.01626892,  0.        ,  0.00889225, -0.09469343,  0.        ,
         0.        , -0.01675144, -0.05828078,  0.02387118, -0.080875  ,
         0.        ,  0.        ]])

Answer: As c become smaller, the coefficient becomes closer to 0

### 23. If you have a logistic regression model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer: 
- use grid search to find the best C param without underfitting
- get more data row
- select more relevent qns for features

---
## Step 6: Answer the problem.

### 24. Suppose you want to understand which psychological features are most important in determining left-handedness. Would you rather use $k$-NN or logistic regression? Why?

Answer: Logistic Regression. It helps to solve binary classification problem, in this scenario is left-handed or right-handed.

### 25. Select your logistic regression model that utilized LASSO regularization with $\alpha = 1$. Interpret the coefficient for `Q1`.

In [52]:
logreg_l1_1.coef_

array([[ 0.02307529, -0.00513949,  0.13479876,  0.02357104,  0.01986972,
        -0.07439096,  0.03440509, -0.15386977, -0.12997462, -0.00688046,
         0.        ,  0.03589063, -0.00486613, -0.06720914, -0.01604056,
         0.06244799,  0.07461695, -0.0575968 ,  0.        , -0.04185905,
        -0.10169811, -0.10451301, -0.11517372, -0.01078743,  0.08977114,
         0.13697412,  0.16718803,  0.07232743,  0.03853853, -0.00344179,
        -0.01765597, -0.02593994,  0.07596326,  0.01336693,  0.17832208,
        -0.05914246,  0.02044956,  0.0341439 , -0.15278653, -0.00563056,
        -0.02276881, -0.06224102, -0.23582888,  0.06533168, -0.10293685,
        -0.004725  , -0.00154582]])

In [53]:
np.exp(logreg_l1_1.coef_)

array([[1.02334359, 0.9948737 , 1.14430648, 1.02385103, 1.02006844,
        0.92830869, 1.03500379, 0.85738367, 0.87811772, 0.99314316,
        1.        , 1.03654247, 0.99514569, 0.93499964, 0.9840874 ,
        1.06443909, 1.07747135, 0.94403051, 1.        , 0.95900494,
        0.90330221, 0.90076308, 0.89121131, 0.98927055, 1.0939239 ,
        1.14679847, 1.18197649, 1.07500728, 1.03929077, 0.99656413,
        0.98249898, 0.97439361, 1.07892294, 1.01345667, 1.19521021,
        0.94257248, 1.02066008, 1.03473349, 0.85831293, 0.99438526,
        0.97748844, 0.93965639, 0.78991583, 1.06751304, 0.90218395,
        0.99528615, 0.99845537]])

Answer: Q1 have 2.3% more odd of predicting if the user is left-handed. 

### 26. If you have to select one model overall to be your *best* model, which model would you select? Why?
- Usually in the "real world," you'll fit many types of models but ultimately need to pick only one! (For example, a client may not understand what it means to have multiple models, or if you're using an algorithm to make a decision, it's probably pretty challenging to use two or more algorithms simultaneously.) It's not always an easy choice, but you'll have to make it soon enough. Pick a model and defend why you picked this model!

Answer: Logistic Regression with Lasso regularisation. Lasso is more suited for categorical variables.

### 27. Circle back to the three specific and conclusively answerable questions you came up with in Q1. Answer one of these for the professor based on the model you selected!

Answer: It is inconclusive as to whether personality, emotional/mental health or intellegence is a good indication of being left-handed or right-handed.

### BONUS:
Looking for more to do? Probably not - you're busy! But if you want to, consider exploring the following. (They could make for a blog post!)
- Create a visual plot comparing training and test metrics for various values of $k$ and various regularization schemes in logistic regression.
- Rather than just evaluating models based on accuracy, consider using sensitivity, specificity, etc.
- In the context of predicting left-handedness, why are unbalanced classes concerning? If you were to re-do this process given those concerns, what changes might you make?
- Fit and evaluate a generalized linear model other than logistic regression (e.g. Poisson regression).
- Suppose this data were in a `SQL` database named `data` and a table named `inventory`. What `SQL` query would return the count of people who were right-handed, left-handed, both, or missing with their class labels of 1, 2, 3, and 0, respectively? (You can assume you've already logged into the database.)